### PR TITLE
[storage][m]: introduce space limitations - fixes #24

### LINKED ISSUE #24: Storage limit for space
@zelima commented on [Wed Mar 21 2018](https://github.com/datahq/specstore/issues/45)

Originally coming from 
https://github.com/datahq/datahub-qa/issues/173
https://github.com/datahq/datahub-qa/issues/174

As an admin I want to limit users ability to publish datasets by space they are using. Eg cannot push more than XXX Mb of data. So that I'm able to offer them storage under different plans

As an admin, I want to limit users ability to publish datasets by private space they are using so that I'm able to offer them privacy under different plans
 
## Acceptance Criteria

* [x] user is getting an appropriate error message via CLI that space limit is reached 
* [x] user is getting an appropriate error message via CLI that private space limit is reached

## Tasks

* [x] Do analysis
* [x] Tests
* [x] add function to models to comunicate with file manager
  * [ ] ~~tests with mocked responses~~
* [x] update controllers to check 

## Analysis

Questions to answer:

* Which number is counted for space usage - raw bytes (data pushed by the user), or total bytes (including derived, original etc...)
* Should we count MBs for each revision?

less important 

* Do private and public bites overlap, Or private bites are counted separately? Eg: I have 100 Mb total free space and published 10 mb of private, how much total space I'm left with 90 or 100?
* What if I don't have public storage, but still have private, can I still publish as public?
* What if I have some MBs left in public and some for private, can I use both to publish as public?

Answering Q1:

1. Assuming users do not really care about the size of derived data, we should only count bytes of the data that user pushes (get's in rawstore). 
2. But on the other hand published dataset (tidy, normalized, various formats, views etc...) is the product we offer. 
besides that, we can't know the size of the remote data. 
Also potentially user can create a bunch of derived data with various processing steps out of the single file.

Considering above I'm assuming we should go with option N2

Answering Q2:

Since a user can maintain N number of revisions as the versions of data think we should count.

### How it's gonna work?

Ok so we already serve API to get the total size of storage for an owner and it's available to filter by findability. That means we can simply hit `api.datahub.io/storage/owner/<owner>?findability=<findability>` from specstore and get the needed info.

```python
# in models
from filemanager.models import get_total_size_for_owner

...
  owner = content['meta']['ownerid']
  findability = content['meta']['findability']
  current_storage = get_total_size_for_owner(owner, findability)(owner, findability if findability == private else None)
  max_storage = limits.get('max_private_storage_mb' if findability == 'private' else 'max_public_storage_mb', 0)
  if max_storage < current_storage:
    errors.append('Max Storage for user exceeded plan limit')
  else:
    continue...
```

### PR DESCRIPTION
**Pull request by zelima**
PR fixes #45

* Uses FileManager API to check the storage limits
* Includes tets

### CODE REVIEW

File: bitstore/blueprint.py [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
-from . import controllers
 
+from . import controllers, services
 
 def make_blueprint():
     """Create blueprint.
     """
 
+    # Create FileManager tables if not exists
+    services.FileRegistry.init_db()
```
Comment (akariv) [inline bitstore/blueprint.py]: I think it's better to do it in one place and not in two - to avoid collision.

Comment (akariv) [reply bitstore/blueprint.py]: (that is - only create the db tables from specstore or here, but not both)

File: bitstore/controllers.py [hunk 2]
```diff
@@ -87,15 +87,36 @@ def authorize(auth_token, req_payload):
         dataset_name = metadata.get('dataset')
         findability = metadata.get('findability')
         acl = 'public-read'
+        is_private = False
         if findability == 'private':
             acl = 'private'
+            is_private = True
```
Comment (akariv) [inline bitstore/controllers.py]: I think it's more better here to use this:
```
is_private = findability == 'private'
acl = 'private' if is_private else 'public-read'
```
shorter and more readable

File: bitstore/controllers.py [hunk 3]
```diff
@@ -87,15 +87,36 @@ def authorize(auth_token, req_payload):
...
+        current_storage = services.FileRegistry.get_total_size_for_owner(
+            owner, 'private' if is_private else None
+        )
+
+        total_bytes = 0
+        for file in req_payload['filedata'].values():
+            total_bytes += file['length']
+
+        bytes = 1000000
+        if current_storage * bytes  + total_bytes >  limit * bytes:
```
Comment (akariv) [inline bitstore/controllers.py]: `get_total_size_for_owner` should return number of bytes, not mbytes...

File: bitstore/services.py [hunk 4]
```diff
@@ -5,6 +5,12 @@
 
 _public_key = None
 
+# FileManager
+from filemanager.models import FileManager
```
Comment (akariv) [inline bitstore/services.py]: why not with the imports?

File: bitstore/services.py [hunk 5]
```diff
@@ -37,6 +43,21 @@ def verify(auth_token, owner):
         return False
 
 
+def permissions(auth_token):
```
Comment (akariv) [inline bitstore/services.py]: please use auth.verifyer for this

Comment (akariv) [reply bitstore/services.py]: Also no real need for _both_ verify and permissions, as they do the same thing...

**Review by akariv — CHANGES_REQUESTED** [review]

File: bitstore/blueprint.py (hunk 1 above)
Comment (zelima) [reply bitstore/blueprint.py]: it'll be only here

File: bitstore/services.py (hunk 4 above)
Comment (zelima) [reply bitstore/services.py]: @akariv not sure what you mean here

File: bitstore/controllers.py [hunk 6]
```diff
@@ -111,8 +108,7 @@ def authorize(auth_token, req_payload):
         for file in req_payload['filedata'].values():
             total_bytes += file['length']
 
-        bytes = 1000000
-        if current_storage * bytes  + total_bytes >  limit * bytes:
+        if current_storage + total_bytes > limit:
```
Comment (akariv) [inline bitstore/controllers.py]: limit is in megabytes :)

File: bitstore/services.py (hunk 4 above)
Comment (akariv) [reply bitstore/services.py]: Why is this import not grouped with the other imports

File: bitstore/services.py [hunk 7]
```diff
@@ -20,59 +21,12 @@ def public_key():
...
 
-def verify(auth_token, owner):
+def verify(auth_token, owner=None):
     """Verify Auth Token.
     :param auth_token: Authentication token to verify
     :param owner: dataset owner
     """
-    if not auth_token:
-        return False
+    verifyer = Verifyer(public_key=public_key())
```
Comment (akariv) [inline bitstore/services.py]: 1. no need to create an object each time.
2. pass the auth endpoint and let Verifyer get the public key from there (and not using the public_key method)

File: bitstore/services.py [hunk 8]
```diff
@@ -20,59 +21,12 @@ def public_key():
...
     if auth_token == 'testing-token' and owner == '__tests':
```
Comment (akariv) [inline bitstore/services.py]: are we using this anywhere? if not, let's remove

**Review by akariv — CHANGES_REQUESTED** [review]

File: bitstore/services.py (hunk 7 above)
Comment (zelima) [reply bitstore/services.py]: @akariv thinking to refactor this as done in specstore (pass it from blueprint) WHYT? 
* https://github.com/datahq/specstore/blob/master/flowmanager/controllers.py#L81-L84
* https://github.com/datahq/specstore/blob/master/flowmanager/blueprint.py#L17-L30

File: bitstore/services.py (hunk 8 above)
Comment (akariv) [reply bitstore/services.py]: so let's do that :)

File: bitstore/services.py (hunk 7 above)
Comment (akariv) [reply bitstore/services.py]: Yes, looks good.

**Review by akariv — APPROVED** [review]
Looks great @zelima !

**Pull request merged**